NPE on DefaultJavaPrettyPrinter#printCtFieldAccess

More intel about bug reproduction can be found on the following issue: https://github.com/HSarah/Tandoori/issues/8

The full stack trace using spoon 5.9.0-SNAPSHOT below
```
spoon.SpoonException: Printing of spoon.support.reflect.code.CtFieldReadImpl on path 
	CtPackageImpl:org
	CtPackageImpl:ethack
	CtPackageImpl:orwall
	CtClassImpl:PreferencesActivity
	CtClassImpl:NetworkPrefs
	CtFieldImpl:listener
	CtNewClassImpl
	CtClassImpl:1
	CtMethodImpl:onSharedPreferenceChanged
	CtBlockImpl
	CtIfImpl
	CtBlockImpl
	CtInvocationImpl
	CtFieldReadImpl:PARAM_ACTIVATE
at position (/home/antoine/Documents/work/dev/sarra/reproduction/orWall/tandoori/sources/org/ethack/orwall/PreferencesActivity.java:126) failed
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.scan(DefaultJavaPrettyPrinter.java:319)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.visitCtInvocation(DefaultJavaPrettyPrinter.java:1177)
	at spoon.support.reflect.code.CtInvocationImpl.accept(CtInvocationImpl.java:57)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.scan(DefaultJavaPrettyPrinter.java:310)
	at spoon.reflect.visitor.printer.ElementPrinterHelper.writeStatement(ElementPrinterHelper.java:139)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.visitCtBlock(DefaultJavaPrettyPrinter.java:491)
	at spoon.support.reflect.code.CtBlockImpl.accept(CtBlockImpl.java:49)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.scan(DefaultJavaPrettyPrinter.java:310)
	at spoon.reflect.visitor.printer.ElementPrinterHelper.writeStatement(ElementPrinterHelper.java:139)
	at spoon.reflect.visitor.printer.ElementPrinterHelper.writeIfOrLoopBlock(ElementPrinterHelper.java:366)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.visitCtIf(DefaultJavaPrettyPrinter.java:1091)
	at spoon.support.reflect.code.CtIfImpl.accept(CtIfImpl.java:46)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.scan(DefaultJavaPrettyPrinter.java:310)
	at spoon.reflect.visitor.printer.ElementPrinterHelper.writeStatement(ElementPrinterHelper.java:139)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.visitCtBlock(DefaultJavaPrettyPrinter.java:491)
	at spoon.support.reflect.code.CtBlockImpl.accept(CtBlockImpl.java:49)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.scan(DefaultJavaPrettyPrinter.java:310)
	at spoon.support.reflect.declaration.CtElementImpl.toString(CtElementImpl.java:249)
	at analyzer.ExecutableProcessor.countEffectiveCodeLines(ExecutableProcessor.java:61)
	at analyzer.ExecutableProcessor.process(ExecutableProcessor.java:42)
	at analyzer.ClassProcessor.processMethods(ClassProcessor.java:66)
	at analyzer.ClassProcessor.process(ClassProcessor.java:53)
	at analyzer.ClassProcessor.process(ClassProcessor.java:25)
	at spoon.support.visitor.ProcessingVisitor.scan(ProcessingVisitor.java:92)
	at spoon.reflect.visitor.CtScanner.visitCtNewClass(CtScanner.java:581)
	at spoon.support.reflect.code.CtNewClassImpl.accept(CtNewClassImpl.java:35)
	at spoon.reflect.visitor.CtScanner.scan(CtScanner.java:150)
	at spoon.support.visitor.ProcessingVisitor.scan(ProcessingVisitor.java:88)
	at spoon.reflect.visitor.CtScanner.visitCtField(CtScanner.java:387)
	at spoon.support.reflect.declaration.CtFieldImpl.accept(CtFieldImpl.java:68)
	at spoon.reflect.visitor.CtScanner.scan(CtScanner.java:150)
	at spoon.support.visitor.ProcessingVisitor.scan(ProcessingVisitor.java:88)
	at spoon.support.visitor.ProcessingVisitor.scan(ProcessingVisitor.java:64)
	at spoon.reflect.visitor.CtScanner.visitCtClass(CtScanner.java:308)
	at spoon.support.reflect.declaration.CtClassImpl.accept(CtClassImpl.java:69)
	at spoon.reflect.visitor.CtScanner.scan(CtScanner.java:150)
	at spoon.support.visitor.ProcessingVisitor.scan(ProcessingVisitor.java:88)
	at spoon.support.visitor.ProcessingVisitor.scan(ProcessingVisitor.java:64)
	at spoon.reflect.visitor.CtScanner.visitCtClass(CtScanner.java:308)
	at spoon.support.reflect.declaration.CtClassImpl.accept(CtClassImpl.java:69)
	at spoon.reflect.visitor.CtScanner.scan(CtScanner.java:150)
	at spoon.support.visitor.ProcessingVisitor.scan(ProcessingVisitor.java:88)
	at spoon.support.visitor.ProcessingVisitor.scan(ProcessingVisitor.java:64)
	at spoon.reflect.visitor.CtScanner.visitCtPackage(CtScanner.java:629)
	at spoon.support.reflect.declaration.CtPackageImpl.accept(CtPackageImpl.java:57)
	at spoon.reflect.visitor.CtScanner.scan(CtScanner.java:150)
	at spoon.support.visitor.ProcessingVisitor.scan(ProcessingVisitor.java:88)
	at spoon.support.visitor.ProcessingVisitor.scan(ProcessingVisitor.java:64)
	at spoon.reflect.visitor.CtScanner.visitCtPackage(CtScanner.java:628)
	at spoon.support.reflect.declaration.CtPackageImpl.accept(CtPackageImpl.java:57)
	at spoon.reflect.visitor.CtScanner.scan(CtScanner.java:150)
	at spoon.support.visitor.ProcessingVisitor.scan(ProcessingVisitor.java:88)
	at spoon.support.visitor.ProcessingVisitor.scan(ProcessingVisitor.java:64)
	at spoon.reflect.visitor.CtScanner.visitCtPackage(CtScanner.java:628)
	at spoon.support.reflect.declaration.CtPackageImpl.accept(CtPackageImpl.java:57)
	at spoon.reflect.visitor.CtScanner.scan(CtScanner.java:150)
	at spoon.support.visitor.ProcessingVisitor.scan(ProcessingVisitor.java:88)
	at spoon.support.visitor.ProcessingVisitor.scan(ProcessingVisitor.java:64)
	at spoon.reflect.visitor.CtScanner.visitCtPackage(CtScanner.java:628)
	at spoon.support.reflect.declaration.CtPackageImpl.accept(CtPackageImpl.java:57)
	at spoon.reflect.visitor.CtScanner.scan(CtScanner.java:150)
	at spoon.support.visitor.ProcessingVisitor.scan(ProcessingVisitor.java:88)
	at spoon.support.QueueProcessingManager.process(QueueProcessingManager.java:119)
	at spoon.support.QueueProcessingManager.process(QueueProcessingManager.java:131)
	at spoon.support.compiler.jdt.JDTBasedSpoonCompiler.process(JDTBasedSpoonCompiler.java:184)
	at spoon.Launcher.process(Launcher.java:707)
	at analyzer.MainProcessor.process(MainProcessor.java:57)
	at analyzer.Main.runAnalysis(Main.java:106)
	at analyzer.Main.main(Main.java:51)
Caused by: java.lang.NullPointerException
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.printCtFieldAccess(DefaultJavaPrettyPrinter.java:790)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.visitCtFieldRead(DefaultJavaPrettyPrinter.java:762)
	at spoon.support.reflect.code.CtFieldReadImpl.accept(CtFieldReadImpl.java:27)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.scan(DefaultJavaPrettyPrinter.java:310)
	... 68 more
```

--- a/src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java
+++ b/src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java
@@ -780,7 +780,7 @@ public class DefaultJavaPrettyPrinter implements CtVisitor, PrettyPrinter {
 				boolean isImportedField = importsContext.isImported(f.getVariable());
 
 				if (!isInitializeStaticFinalField && !(isStaticField && isImportedField)) {
-					if (target.isImplicit()) {
+					if (target.isImplicit() && !(f.getVariable().getFieldDeclaration() == null && this.env.getNoClasspath())) {
 						/*
 						 * target is implicit, check whether there is no conflict with an local variable, catch variable or parameter
 						 * in case of conflict make it explicit, otherwise the field access is shadowed by that variable.

--- a/src/test/java/spoon/test/prettyprinter/DefaultPrettyPrinterTest.java
+++ b/src/test/java/spoon/test/prettyprinter/DefaultPrettyPrinterTest.java
@@ -30,6 +30,7 @@ import java.util.Arrays;
 import java.util.List;
 
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
 import static spoon.testing.utils.ModelUtils.build;
 
@@ -292,6 +293,18 @@ public class DefaultPrettyPrinterTest {
 		// Pretty-printing the Spooned code snippet and compiling the resulting code.
 		snippet = launcher.getFactory().Code().createCodeSnippetStatement(compile.toString());
 		assertEquals(compile, snippet.compile());
+	}
+
+	@Test
+	public void testIssue1501() {
+		Launcher launcher = new Launcher();
+		launcher.addInputResource("./src/test/resources/noclasspath/orwall/PreferencesActivity.java");
+		launcher.addInputResource("./src/test/resources/noclasspath/orwall/BackgroundProcess.java");
+		launcher.setSourceOutputDirectory("./target/issue1501");
+		launcher.getEnvironment().setNoClasspath(true);
+		launcher.run();
+
+		assertFalse(launcher.getModel().getAllTypes().isEmpty());
 	}
 
 }

--- a/src/test/java/spoon/test/prettyprinter/PrinterTest.java
+++ b/src/test/java/spoon/test/prettyprinter/PrinterTest.java
@@ -1,12 +1,14 @@
 package spoon.test.prettyprinter;
 
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
 import static spoon.testing.utils.ModelUtils.canBeBuilt;
 
 import org.junit.Test;
 
+import org.mockito.internal.matchers.Null;
 import spoon.Launcher;
 import spoon.SpoonException;
 import spoon.compiler.SpoonResourceHelper;
@@ -155,6 +157,7 @@ public class PrinterTest {
 			type.getMethodsByName("failingMethod").get(0).getBody().getStatement(0).toString();
 			fail();
 		} catch (SpoonException e) {
+			assertTrue(e.getCause() instanceof NullPointerException);
 			//the name of the missing field declaration is part of exception
 			assertTrue(e.getMessage().indexOf("testedField")>=0);
 			//the name of the method where field declaration is missing is part of exception

--- a/src/test/resources/noclasspath/orwall/BackgroundProcess.java
+++ b/src/test/resources/noclasspath/orwall/BackgroundProcess.java
@@ -1,0 +1,32 @@
+package org.ethack.orwall;
+
+import android.app.IntentService;
+import android.content.Intent;
+
+import org.ethack.orwall.iptables.InitializeIptables;
+import org.ethack.orwall.lib.Constants;
+
+/**
+ * Created by cedric on 7/31/14.
+ */
+public class BackgroundProcess extends IntentService {
+
+    public BackgroundProcess() {
+        super("BackroundProcess");
+    }
+
+    @Override
+    protected void onHandleIntent(Intent workIntent) {
+        String action = workIntent.getStringExtra(Constants.ACTION);
+
+        if (action.equals(Constants.ACTION_PORTAL)) {
+            boolean activate = workIntent.getBooleanExtra(Constants.PARAM_ACTIVATE, false);
+            managePortal(activate);
+        }
+    }
+
+    private void managePortal(boolean activate) {
+        InitializeIptables initializeIptables = new InitializeIptables(this);
+        initializeIptables.enableCaptiveDetection(activate, this);
+    }
+}

--- a/src/test/resources/noclasspath/orwall/PreferencesActivity.java
+++ b/src/test/resources/noclasspath/orwall/PreferencesActivity.java
@@ -1,0 +1,161 @@
+package org.ethack.orwall;
+
+import android.content.Context;
+import android.content.Intent;
+import android.content.SharedPreferences;
+import android.os.Bundle;
+import android.preference.PreferenceActivity;
+import android.preference.PreferenceFragment;
+import android.preference.PreferenceManager;
+
+import org.ethack.orwall.iptables.InitializeIptables;
+
+import java.util.Arrays;
+import java.util.List;
+
+
+public class PreferencesActivity extends PreferenceActivity {
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+    }
+
+    @Override
+    public void onBuildHeaders(List<Header> target) {
+        loadHeadersFromResource(R.xml.preferences_header, target);
+    }
+
+    @Override
+    protected boolean isValidFragment(String fragmentName) {
+        String prepend = "org.ethack.orwall.PreferencesActivity$";
+        String[] fragments = {
+                prepend + "ScriptPrefs",
+                prepend + "SpecialApps",
+                prepend + "NetworkPrefs",
+                prepend + "ProxyPorts",
+        };
+
+        return Arrays.asList(fragments).contains(fragmentName);
+    }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+    }
+
+    public static class ScriptPrefs extends PreferenceFragment {
+
+        private SharedPreferences.OnSharedPreferenceChangeListener listener = new SharedPreferences.OnSharedPreferenceChangeListener() {
+            @Override
+            public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String s) {
+                InitializeIptables iptables = new InitializeIptables(getActivity());
+                if (sharedPreferences.getBoolean(s, true) && s.equals("enforce_init_script")) {
+                    iptables.installInitScript(getActivity());
+                }
+                if (sharedPreferences.getBoolean(s, true) && s.equals("deactivate_init_script") && !sharedPreferences.getBoolean("enforce_init_script", true)) {
+                    iptables.removeIniScript();
+                }
+            }
+        };
+
+        @Override
+        public void onCreate(Bundle savedInstanceState) {
+            super.onCreate(savedInstanceState);
+
+            PreferenceManager.setDefaultValues(getActivity(), R.xml.network_preference, true);
+            addPreferencesFromResource(R.xml.fragment_init_pref);
+        }
+
+        @Override
+        public void onResume() {
+            super.onResume();
+            getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(listener);
+        }
+
+        @Override
+        public void onPause() {
+            super.onPause();
+            getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(listener);
+        }
+    }
+
+    public static class SpecialApps extends PreferenceFragment {
+        private SharedPreferences.OnSharedPreferenceChangeListener listener = new SharedPreferences.OnSharedPreferenceChangeListener() {
+            @Override
+            public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String s) {
+                return;
+            }
+        };
+
+        @Override
+        public void onCreate(Bundle savedInstanceState) {
+            super.onCreate(savedInstanceState);
+            //PreferenceManager.setDefaultValues(getActivity(), R.xml.fragment_apps_prefs, true);
+            addPreferencesFromResource(R.xml.fragment_apps_prefs);
+        }
+
+        @Override
+        public void onResume() {
+            super.onResume();
+            getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(listener);
+        }
+
+        @Override
+        public void onPause() {
+            super.onPause();
+            getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(listener);
+        }
+    }
+
+    public static class NetworkPrefs extends PreferenceFragment {
+        private SharedPreferences.OnSharedPreferenceChangeListener listener = new SharedPreferences.OnSharedPreferenceChangeListener() {
+            @Override
+            public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String s) {
+                InitializeIptables iptables = new InitializeIptables(getActivity());
+                if (s.equals("enable_lan")) {
+                    iptables.LANPolicy(sharedPreferences.getBoolean(s, false));
+                }
+
+                if (s.equals("enable_adb")) {
+                    iptables.enableADB(sharedPreferences.getBoolean(s, false));
+                }
+
+                if (s.equals("enable_captive_portal")) {
+                    Context context = getActivity();
+                    Intent bgpProcess = new Intent(context, BackgroundProcess.class);
+                    bgpProcess.putExtra(BackgroundProcess.PARAM_ACTIVATE, sharedPreferences.getBoolean(s, false));
+                    bgpProcess.putExtra(BackgroundProcess.ACTION, BackgroundProcess.ACTION_PORTAL);
+                    context.startService(bgpProcess);
+                }
+            }
+        };
+
+        @Override
+        public void onCreate(Bundle savedInstanceState) {
+            super.onCreate(savedInstanceState);
+            PreferenceManager.setDefaultValues(getActivity(), R.xml.other_preferences, true);
+            addPreferencesFromResource(R.xml.fragment_network_prefs);
+        }
+
+        @Override
+        public void onResume() {
+            super.onResume();
+            getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(listener);
+        }
+
+        @Override
+        public void onPause() {
+            super.onPause();
+            getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(listener);
+        }
+    }
+
+    public static class ProxyPorts extends PreferenceFragment {
+        @Override
+        public void onCreate(Bundle savedInstanceState) {
+            super.onCreate(savedInstanceState);
+            addPreferencesFromResource(R.xml.fragment_proxy_ports);
+        }
+    }
+
+}

